### PR TITLE
Make watch order conformance test reliable

### LINKED ISSUE #101949: "should receive events on concurrent watches in same order" conformance test is unreliable
#### What happened:

By inspection, the "should receive events on concurrent watches in same order" conformance test does two things that are not stable:
* Starts watches from resourceVersion=0
* Relies on watch connections without handling timeouts/restarts

This test can fail on clusters that have configmaps autocreated in namespaces already because ordering of watch events for resourceVersion=0 watches is undefined (see https://github.com/kubernetes/kubernetes/issues/74022)

#### What you expected to happen:

Watch connections used by conformance tests should use well-defined ordering (starting from resourceVersions other than 0) and handle interruptions properly.

#### Environment:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind bug
/kind failing-test

#### What this PR does / why we need it:

Fixes https://github.com/kubernetes/kubernetes/issues/101949

* Starts the watch order conformance test from a non-zero resourceVersion (required to get consistent ordering in a namespace with multiple watch events already present for configmaps)
* Handles dropped watches with a retry watcher

```release-note
Fixes the `should receive events on concurrent watches in same order` conformance test to work properly on clusters that auto-create additional configmaps in namespaces
```

### CODE REVIEW

**Comment by liggitt** [comment]
test failure is #101953, unrelated

/retest

File: test/e2e/apimachinery/watch.go [hunk 1]
```diff
@@ -436,11 +448,14 @@ func waitForEvent(w watch.Interface, expectType watch.EventType, expectObject ru
 
 func waitForNextConfigMapEvent(watch watch.Interface) *v1.ConfigMap {
 	select {
-	case event := <-watch.ResultChan():
+	case event, ok := <-watch.ResultChan():
+		if !ok {
+			framework.Failf("Watch closed unexpectedly")
```
[the remark this answers is not in the document]
Comment (liggitt) [reply test/e2e/apimachinery/watch.go:453]: This is a retry watcher, so it is already retrying/re-establishing internally... we only get to this point if the retry failed. Still possible, but more robust than it was before.

[the remark this answers is not in the document]
Comment (dims) [reply test/e2e/apimachinery/watch.go:453]: Nice!

**Comment by liggitt** [comment]
opened picks to 1.19, 1.20, 1.21

**Pull request merged**